src/core/Akka.Streams/Dsl/Hub.cs

```diff
@@ -1169,4 +1169,707 @@ public BroadcastHub(int bufferSize)
+            public static ConsumerQueue Empty { get; } = new ConsumerQueue(ImmutableQueue<object>.Empty, 0);
+
+            private readonly ImmutableQueue<object> _queue;
+
+            public ConsumerQueue(ImmutableQueue<object> queue, int size)
+            {
+                _queue = queue;
+                Size = size;
+            }
+
+            public ConsumerQueue Enqueue(object element) => new ConsumerQueue(_queue.Enqueue(element), Size + 1);
+
+            public bool IsEmpty => Size == 0;
+
+            public object Head => _queue.First();
+
+            public ConsumerQueue Tail => new ConsumerQueue(_queue.Dequeue(), Size - 1);
+
+            public int Size { get; }
+        }
+
+        private sealed class PartitionQueue : IPartitionQueue
+        {
+            private AtomicCounter _totalSize = new AtomicCounter();
+            private readonly ConcurrentDictionary<long, ConsumerQueue> _queues = new ConcurrentDictionary<long, ConsumerQueue>();
+
+            public void Init(long id) => _queues.TryAdd(id, ConsumerQueue.Empty);
+
+            public int TotalSize => _totalSize.Current;
+
+            public int Size(long id)
+            {
+                if (_queues.TryGetValue(id, out var queue))
+                    return queue.Size;
+
+                throw new ArgumentException($"Invalid stream identifier: {id}", nameof(id));
+            }
+
+            public bool IsEmpty(long id)
+            {
+                if (_queues.TryGetValue(id, out var queue))
+                    return queue.IsEmpty;
+
+                throw new ArgumentException($"Invalid stream identifier: {id}", nameof(id));
+            }
+
+            public bool NonEmpty(long id) => !IsEmpty(id);
+
+            public void Offer(long id, object element)
+            {
+                if (_queues.TryGetValue(id, out var queue))
+                {
+                    if (_queues.TryUpdate(id, queue.Enqueue(element), queue))
+                        _totalSize.IncrementAndGet();
+                    else
+                        Offer(id, element);
+                }
+                else
+                    throw new ArgumentException($"Invalid stream identifier: {id}", nameof(id));
+            }
+
+            public object Poll(long id)
+            {
+                var success = _queues.TryGetValue(id, out var queue);
+                if (!success || queue.IsEmpty)
+                    return null;
+
+                if (_queues.TryUpdate(id, queue.Tail, queue))
+                {
+                    _totalSize.Decrement();
+                    return queue.Head;
+                }
+
+                return Poll(id);
+            }
+
+            public void Remove(long id)
+            {
+                if (_queues.TryRemove(id, out var queue))
+                    _totalSize.AddAndGet(-queue.Size);
+            }
+        }
+
+        #endregion
+
+        #region internal classes
+
+        private interface IConsumerEvent { }
```

Comment:
I don't consider this a mistake, but I'm starting to think, that in the future, we should favor tagged unions over this approach - no in every case, but for sure in cases like several singletons and one type with a error field. Example:

```csharp
struct ConsumerEvent 
{
    enum ConsumerEventType { Wakeup, Initialize, HubCompleted }
    
    readonly ConsumerEventType Type;
    readonly Exception Failure;
    ....
}
```